app/src/main/java/de/danoeh/antennapod/activity/SingleFileAddActivity.java

```diff
@@ -0,0 +1,220 @@
+                singleFileExplanation1.setVisibility(View.GONE);
+                findViewById(R.id.divider1).setVisibility(View.GONE);
+                butChooseFilesystem.setVisibility(View.GONE);
+            }
+        }
+        if(singleFileExplanation1.getVisibility() == View.VISIBLE) {
+            singleFileHeaderExplanation1.setText(String.format(optionLabel, nextOption));
+            nextOption++;
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        StorageUtils.checkStorageAvailability(this);
+    }
+
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
+            case android.R.id.home:
+                finish();
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    /*
+     * Creates an implicit intent to launch a file manager which lets
+     * the user choose a specific file to import.
+     */
+    private void chooseFileFromFilesystem() {
+        int permission = ActivityCompat.checkSelfPermission(this, android.Manifest.permission.READ_EXTERNAL_STORAGE);
+        if (permission != PackageManager.PERMISSION_GRANTED) {
+            requestPermission();
+        }
+
+        try {
+            startActivityForResult(intentPickAction, CHOOSE_SINGLE_FILE);
+        } catch (ActivityNotFoundException e) {
+            Log.e(TAG, "No activity found. Should never happen...");
+        }
+    }
+
+    /**
+      * Gets the path of the file chosen with chooseFileToImport()
+      */
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        //super.onActivityResult(requestCode, resultCode, data);
+        if (resultCode == RESULT_OK && requestCode == CHOOSE_SINGLE_FILE) {
+            Uri uri = data.getData();
+            if(uri != null && uri.toString().startsWith("/")) {
+                uri = Uri.parse("file://" + uri.toString());
+            }
+            boolean ret = importUri(uri);
+            if (ret == false) {
+                Toast.makeText(this, "Could not import", Toast.LENGTH_LONG).show();
+            }
+
+            //go back to main
+            Intent intent = new Intent(this, MainActivity.class);
+            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
+                    | Intent.FLAG_ACTIVITY_NEW_TASK);
+            startActivity(intent);
+        } else {
+            Log.d(TAG, "result not ok or code not file: " + resultCode + "," + requestCode);
+        }
+    }
+
+    private void requestPermission() {
+        String[] permissions = { android.Manifest.permission.READ_EXTERNAL_STORAGE };
+        ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_READ_EXTERNAL_STORAGE);
+    }
```

Comment:
There is a callback when the permission was granted. With the callback, it is possible to retry to pick a folder directly.

```
@Override
public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
    if (requestCode == PERMISSION_REQUEST_READ_EXTERNAL_STORAGE) {
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            chooseFileFromFilesystem(); // Retry
        }
    }
}
```